### PR TITLE
feat(expect): allow chaining expects

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: tests/playwright-test/expect.spec.ts [hunk 1]
```diff
@@ -704,3 +709,155 @@ test('should not leak long expect message strings', async ({ runInlineTest }) =>
...
+              \`Expected: \${this.isNot ? 'not' : ''}\${this.utils.printExpected(expected)}\\n\` +
+              (matcherResult ? \`Received: \${this.utils.printReceived(matcherResult.actual)}\` : '') +
+              log
+          : () =>  this.utils.matcherHint('toBe', locator, expected, expectOptions) +
+              '\\n\\n' +
+              \`Expected: \${this.utils.printExpected(expected)}\n\` +
+              (matcherResult ? \`Received: \${this.utils.printReceived(matcherResult.actual)}\` : '') +
+              log;
+
+        return {
```
Comment (dgozman) [inline tests/playwright-test/expect.spec.ts:837]: If I don't return all these fields in my custom matcher, will things keep working?

File: utils/generate_types/overrides-test.d.ts [hunk 2]
```diff
@@ -341,38 +341,92 @@ type AllMatchers<R, T> = PageAssertions & LocatorAssertions & APIResponseAsserti
...
+  <T = unknown>(actual: T, messageOrOptions?: string | { message?: string }): MakeMatchers<void, T, ExtendedMatchers>;
+  soft: <T = unknown>(actual: T, messageOrOptions?: string | { message?: string }) => MakeMatchers<void, T, ExtendedMatchers>;
   poll: <T = unknown>(actual: () => T | Promise<T>, messageOrOptions?: string | { message?: string, timeout?: number, intervals?: number[] }) => BaseMatchers<Promise<void>, T> & {
     /**
      * If you know how to test something, `.not` lets you test its opposite.
      */
      not: BaseMatchers<Promise<void>, T>;
   };
-  extend(matchers: any): void;
+  extend<MoreMatchers extends Record<string, (this: State, receiver: unknown, ...args: any[]) => MatcherReturnType | Promise<MatcherReturnType>>>(matchers: MoreMatchers): Expect<ExtendedMatchers & MoreMatchers>;
```
Comment (dgozman) [inline utils/generate_types/overrides-test.d.ts:424]: Judging by our own `toHaveLoggedSoftwareDownload`, forcing the return type is a breaking change.

File: utils/generate_types/overrides-test.d.ts [hunk 3]
```diff
@@ -341,38 +341,92 @@ type AllMatchers<R, T> = PageAssertions & LocatorAssertions & APIResponseAsserti
...
+  expectedColor?: MatcherHintColor;
+  isDirectExpectCall?: boolean;
+  isNot?: boolean;
+  promise?: string;
+  receivedColor?: MatcherHintColor;
+  secondArgument?: string;
+  secondArgumentColor?: MatcherHintColor;
+};
+
+export interface JestMatcherUtils {
```
Comment (dgozman) [inline utils/generate_types/overrides-test.d.ts:363]: Are you sure we want to expose all of this?

File: utils/generate_types/overrides-test.d.ts [hunk 4]
```diff
@@ -341,38 +341,92 @@ type AllMatchers<R, T> = PageAssertions & LocatorAssertions & APIResponseAsserti
 
 type IfAny<T, Y, N> = 0 extends (1 & T) ? Y : N;
 type Awaited<T> = T extends PromiseLike<infer U> ? U : T;
-type MakeMatchers<R, T> = {
+type ToUserMatcher<F> = F extends (a: any, ...args: infer Rest) => infer R ? (...args: Rest) => (R extends PromiseLike<infer U> ? Promise<void> : void) : never;
+type ToUserMatcherForArgType<T, ArgType> = T extends (arg: ArgType, ...rest: any[]) => any ? ToUserMatcher<T> : never;
```
Comment (dgozman) [inline utils/generate_types/overrides-test.d.ts:345]: I'd inline `ToUserMatcher` here:

```suggestion
type ToUserMatcherForArgType<T, ArgType> = T extends (arg: ArgType, ...rest: infer Rest) => infer R ? (...args: Rest) => (R extends PromiseLike<infer U> ? Promise<void> : void) : never;
```

File: tests/playwright-test/expect.spec.ts (hunk 1 above)
Comment (pavelfeldman) [reply tests/playwright-test/expect.spec.ts:837]: You must return `message` and `pass`. The rest is optional.

File: utils/generate_types/overrides-test.d.ts (hunk 3 above)
Comment (pavelfeldman) [reply utils/generate_types/overrides-test.d.ts:363]: I definitely don't, but I think this ship has sailed.

File: utils/generate_types/overrides-test.d.ts (hunk 2 above)
Comment (pavelfeldman) [reply utils/generate_types/overrides-test.d.ts:424]: I think toHaveLoggedSoftwareDownload is not right, i'll fix it.

**Review by dgozman — APPROVED** [review]

**Pull request merged**